demo-circuit/src/blaze_csw/constraints/data_structures.rs

```diff
@@ -0,0 +1,1821 @@
+        _cs: CS,
+        _f: F,
+    ) -> Result<Self, SynthesisError>
+    where
+        F: FnOnce() -> Result<T, SynthesisError>,
+        T: Borrow<CswFtOutputData>,
+    {
+        unimplemented!()
+    }
+}
+
+impl ConstantGadget<CswFtOutputData, FieldElement> for CswFtOutputDataGadget {
+    fn from_value<CS: ConstraintSystemAbstract<FieldElement>>(
+        _cs: CS,
+        value: &CswFtOutputData,
+    ) -> Self {
+        let amount_g = UInt64::constant(value.amount);
+        let receiver_pub_key_g = bytes_to_bits(&value.receiver_pub_key)
+            .iter()
+            .map(|&bit| Boolean::constant(bit))
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
+        let payback_addr_data_hash_g = bytes_to_bits(&value.payback_addr_data_hash)
+            .iter()
+            .map(|&bit| Boolean::constant(bit))
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
+        let tx_hash_g = bytes_to_bits(&value.tx_hash)
+            .iter()
+            .map(|&bit| Boolean::constant(bit))
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
+        let out_idx_g = UInt32::constant(value.out_idx);
+
+        Self {
+            amount_g,
+            receiver_pub_key_g,
+            payback_addr_data_hash_g,
+            tx_hash_g,
+            out_idx_g,
+        }
+    }
+
+    fn get_constant(&self) -> CswFtOutputData {
+        unimplemented!();
+    }
+}
+
+impl EqGadget<FieldElement> for CswFtOutputDataGadget {
+    fn is_eq<CS: ConstraintSystemAbstract<FieldElement>>(
+        &self,
+        mut cs: CS,
+        other: &Self,
+    ) -> Result<Boolean, SynthesisError> {
+        let b1 = self
+            .amount_g
+            .is_eq(cs.ns(|| "is eq amount_g"), &other.amount_g)?;
+        let b2 = self.receiver_pub_key_g.is_eq(
+            cs.ns(|| "is eq receiver_pub_key_g"),
+            &other.receiver_pub_key_g,
+        )?;
+        let b3 = self.payback_addr_data_hash_g.is_eq(
+            cs.ns(|| "is eq payback_addr_data_hash_g"),
+            &other.payback_addr_data_hash_g,
+        )?;
+        let b4 = self
+            .tx_hash_g
+            .is_eq(cs.ns(|| "is eq tx_hash_g"), &other.tx_hash_g)?;
+        let b5 = self
+            .out_idx_g
+            .is_eq(cs.ns(|| "is eq out_idx_g"), &other.out_idx_g)?;
+
+        Boolean::kary_and(
+            cs.ns(|| "is_eq CswUtxoInputDataGadget"),
+            &[b1, b2, b3, b4, b5],
+        )
+    }
+}
+
+impl ToConstraintFieldGadget<FieldElement> for CswFtOutputDataGadget {
+    type FieldGadget = FieldElementGadget;
+
+    fn to_field_gadget_elements<CS: ConstraintSystemAbstract<FieldElement>>(
+        &self,
+        mut cs: CS,
+    ) -> Result<Vec<FieldElementGadget>, SynthesisError> {
+        let mut bits = self.amount_g.to_bits_le();
+        bits.reverse();
+
+        let mut receiver_pub_key_bits = self.receiver_pub_key_g;
```

Comment:
There is a misalignment between how mainchain use `receiver_pub_key` for ForwardTransfers and how it is used for proof of public key ownership in `ScPublicKeyGadget::enforce_pk_ownership` inside CSW circuit.

On MC: to send a FT to sidechain we need to provide MC RPC with sc public key hex string generated on SC node.
This public key bytes are transformed into the hex string exactly as it is used by SC noded. Instead, MC reverse that value under the hood, because it is represented as a `uint256`. And this reversed representation is used while computing CommitmentTree.
So, on SDK side pub key is "aabbccdd", on MC is "ddccbbaa".

Then when we want to withdraw FT in a CSW we need to provide FT data exactly how MC/CCTP lib expect it for commitment tree computation. So, to prove that some FT exists in some CommTree `pubKeyBytes` must be passed as "ddccbbaa".

From other side CSW circuit checks the ownership of the public key, but expect it in the original form "aabbccdd".
Because the "receiver_pub_key_g" is a single source of data for both of these checks we will not be able to satisfy both of them inside the circuit.
Note: there is no problem with UTXO withdrawals, since `pubKeyBytes` are used in the expected order.

I would propose to use `pubKeyBytes` in the original order when passing as the witness of CSW circuit and when calculating the nullifier outside the circuit, but reverse the bytes here before calculating the `ft_hash`.

Reply:
Done